### PR TITLE
Add linter wastedassign

### LINKED ISSUE #1492: Support for wastedassign
Hi :)

[wastedassign](https://github.com/sanposhiho/wastedassign)

> wastedassign finds wasted assignment statements
> found the value ...
> - reassigned, but never used afterwards
> - reassigned, but reassigned  without using value

Go compiler can find wasted assign, but cannot find wasted reassign.

#### Example

```
func f() int {
	a := 0 
        b := 0
        fmt.Print(a)
        fmt.Print(b)
        a = 1  // This reassignment is wasted, because never used afterwards. Wastedassign find this 

        b = 1  // This reassignment is wasted, because reassigned without using this value. Wastedassign find this 
        b = 2
        fmt.Print(b)
        
	return 1 + 2
}
```

### PR DESCRIPTION
**Pull request by sanposhiho**
closes #1492

Hi :D

Add [wastedassign](https://github.com/sanposhiho/wastedassign)

`wastedassign` finds wasted assignment statements

found the value ...

- reassigned, but never used afterward
- reassigned, but reassigned without using the value

```go
func f() int {
	a := 0 
        b := 0
        fmt.Print(a)
        fmt.Print(b)
        a = 1  // This reassignment is wasted, because never used afterwards. Wastedassign find this 

        b = 1  // This reassignment is wasted, because reassigned without use this value. Wastedassign find this 
        b = 2
        fmt.Print(b)
        
	return 1 + 2
}
```

### CODE REVIEW

**Comment by ldez** [comment]
Hello @sanposhiho,

could you explain why you created a fork of `golang.org/x/tools` (https://github.com/sanposhiho/tools) ?

I see your change:

https://github.com/sanposhiho/tools/blob/da3530a453b58298252f9d97b5419d0fc231f00b/go/analysis/passes/buildssa/buildssa.go#L51
```go
mode := ssa.BuilderMode(ssa.NaiveForm)
```

https://github.com/golang/tools/blob/e0d201561e39aeaab8136532f2bbe17ae1ef24ff/go/analysis/passes/buildssa/buildssa.go#L49-L51
```go
	// Some Analyzers may need GlobalDebug, in which case we'll have
	// to set it globally, but let's wait till we need it.
	mode := ssa.BuilderMode(0)
```

It seems, when I read the comment, that a PR on that point on the repo https://github.com/golang/tools will be welcome.

I pay attention to this point because it can impact golangci-lint's performance because this lib is at the heart of golangci-lint's performance optimizations.

**Comment by sanposhiho** [comment]
Hi @ldez!

> could you explain why you created a fork of golang.org/x/tools

I use the mode `ssa.NaiveForm`, because without this mode, the process called "lifting" replaces the store process with ssa.Value and can't track the use of variables anymore.

> It seems, when I read the comment, that a PR on that point on the repo https://github.com/golang/tools will be welcome.
I pay attention to this point because it can impact golangci-lint's performance because this lib is at the heart of golangci-lint's performance optimizations.

OK, I'll give it a PR once and see how it goes.
If the PR is approved, I'll make the changes to use golang.org/x/tools.

**Comment by sanposhiho** [comment]
Hi team.
I'm sorry it took me so long to reply.

I removed the dependency of wastedassign on sanposhiho/tools, as it seemed to take a while for the proposal to go/tools to be supported.

File: go.mod [hunk 1]
```diff
@@ -54,7 +54,8 @@ require (
 	github.com/polyfloyd/go-errorlint v0.0.0-20201127212506-19bd8db6546f
 	github.com/ryancurrah/gomodguard v1.2.0
 	github.com/ryanrolds/sqlclosecheck v0.3.0
-	github.com/securego/gosec/v2 v2.6.1
+	github.com/sanposhiho/wastedassign v0.1.2
+	github.com/securego/gosec/v2 v2.5.0
```
Comment (SVilgelm) [inline go.mod:58]: hmm, please revert the changes for `gosec`

**Review by SVilgelm — CHANGES_REQUESTED** [review]

File: go.mod (hunk 1 above)
Comment (sanposhiho) [reply go.mod]: Oops, sorry.

**Comment by SVilgelm** [comment]
it would be great if you can use `golang.org/x/tools v0.1.0` instead of `v0.0.0-20200918232735-d647fc253266`, just to be consistent with `golangci-lint`

**Review by ldez — CHANGES_REQUESTED** [review]
Could you replace the panic by errors?

https://github.com/sanposhiho/wastedassign/blob/master/wastedassign.go#L75
https://github.com/sanposhiho/wastedassign/blob/master/wastedassign.go#L80

**Comment by sanposhiho** [comment]
@ldez   @SVilgelm 
Thanks for reviews, fixed them

File: pkg/lint/lintersdb/manager.go [hunk 2]
```diff
@@ -374,6 +374,10 @@ func (m Manager) GetAllSupportedLinterConfigs() []*linter.Config {
 			WithPresets(linter.PresetBugs).
 			WithLoadForGoAnalysis().
 			WithURL("https://github.com/charithe/durationcheck"),
+		linter.NewConfig(golinters.NewWastedAssign()).
+			WithLoadForGoAnalysis().
+			WithPresets(linter.PresetStyle).
+			WithURL("https://github.com/sanposhiho/wastedassign"),
```
Comment (ldez) [inline pkg/lint/lintersdb/manager.go:380]:
```suggestion
		linter.NewConfig(golinters.NewWastedAssign()).
			WithPresets(linter.PresetStyle).
			WithLoadForGoAnalysis().
			WithURL("https://github.com/sanposhiho/wastedassign"),
```

**Review by SVilgelm — APPROVED** [review]
Looks good to me. Thank you

**Review by ldez — APPROVED** [review]
LGTM

**Comment by stefanvanburen** [comment]
I'm curious about the difference between [wastedassign](https://github.com/sanposhiho/wastedassign) (https://golangci-lint.run/docs/linters/configuration/#wastedassign) and [ineffassign](https://github.com/gordonklaus/ineffassign) (https://golangci-lint.run/docs/linters/configuration/#ineffassign); this was [asked upstream in wastedassign](https://github.com/sanposhiho/wastedassign/issues/42) but there hasn't been an answer in 2 years.

Checking an analyzer run, `wastedassign` is one of the highest time usage behind building, so it's tempting to disable it in favor of `ineffeassign`.

```
INFO [linters_context/goanalysis] analyzers took 21m45.544450121s with top 10 stages: buildir: 3m0.931518922s, wastedassign: 1m11.074647138s, gosec: 27.541418458s,
```
